rocAL/source/readers/file_source_reader.cpp

```diff
@@ -60,30 +83,38 @@ Reader::Status FileSourceReader::initialize(ReaderConfig desc) {
     _loop = desc.loop();
     _meta_data_reader = desc.meta_data_reader();
     _last_batch_info = desc.get_last_batch_policy();
+    _pad_last_batch_repeated = _last_batch_info.second;
+    _stick_to_shard = desc.get_stick_to_shard();
+    _shard_size = desc.get_shard_size();
     ret = subfolder_reading();
+    _curr_file_idx = get_start_idx();  // shard's start_idx would vary for every shard in the vector
     // shuffle dataset if set
     if (ret == Reader::Status::OK && _shuffle)
-        std::random_shuffle(_file_names.begin(), _file_names.end());
+        std::random_shuffle(_file_names.begin() + _curr_file_idx,
+                            _file_names.begin() + _curr_file_idx + actual_shard_size_without_padding());
 
     return ret;
 }
 
+void FileSourceReader::increment_curr_file_idx() {
+    // The condition satisfies for both pad_last_batch = True (or) False
+    auto start_idx_of_shard = get_start_idx();
+    if (_stick_to_shard == false) {  // The elements of each shard rotate in a round-robin fashion once the elements in particular shard is exhausted
+        _curr_file_idx = (_curr_file_idx + 1) % _file_names.size();
+    } else {  // Stick to only elements from the current shard
+        if (_curr_file_idx >= start_idx_of_shard &&
```

Comment:
use precomputed start and end instead

Reply:
the start idx and end idx can differ once we move to next shards data when stick_to_shard = False.
Precomputed the start and end at the beginning of calculating the curr_file_idx.